### PR TITLE
ADBDEV-4944: Add assertions to return from dynamic_cast.

### PR DESCRIPTION
**Pull request by RekGRpth**
Add assertions to return from dynamic_cast.

dynamic_cast may return NULL, so I added assertions.

### CODE REVIEW

File: src/backend/gporca/libnaucrates/src/parser/CParseHandlerAgg.cpp [hunk 1]
```diff
@@ -167,11 +168,15 @@ CParseHandlerAgg::EndElement(const XMLCh *const,  // element_uri,
...
+	GPOS_ASSERT(NULL != prop_parse_handler);
 	CParseHandlerUtils::SetProperties(m_dxl_node, prop_parse_handler);
 
 	// add children
+	GPOS_ASSERT(NULL != proj_list_parse_handler);
 	AddChildFromParseHandler(proj_list_parse_handler);
+	GPOS_ASSERT(NULL != filter_parse_handler);
 	AddChildFromParseHandler(filter_parse_handler);
+	GPOS_ASSERT(NULL != child_parse_handler);
 	AddChildFromParseHandler(child_parse_handler);
```
Comment (KnightMurloc) [inline src/backend/gporca/libnaucrates/src/parser/CParseHandlerAgg.cpp]: Are these changes really necessary? The original task only talks about proj_list_parse_handler.
upd:
The question also applies to the remaining modified files.

Comment (RekGRpth) [reply src/backend/gporca/libnaucrates/src/parser/CParseHandlerAgg.cpp]: These changes are needed for other numbers, for example, proj_list_parse_handler for 99.

**Comment by KnightMurloc** [comment]
Maybe we should do the checks in such a way that there are fewer changes?
```c
	CParseHandlerProperties *prop_parse_handler =
		dynamic_cast<CParseHandlerProperties *>((*this)[0]);
	CParseHandlerGroupingColList *grouping_col_list_parse_handler =
		dynamic_cast<CParseHandlerGroupingColList *>((*this)[1]);
	CParseHandlerProjList *proj_list_parse_handler =
		dynamic_cast<CParseHandlerProjList *>((*this)[2]);
	CParseHandlerFilter *filter_parse_handler =
		dynamic_cast<CParseHandlerFilter *>((*this)[3]);
	CParseHandlerPhysicalOp *child_parse_handler =
		dynamic_cast<CParseHandlerPhysicalOp *>((*this)[4]);
	GPOS_ASSERT(prop_parse_handler &&
		grouping_col_list_parse_handler &&
		proj_list_parse_handler &&
		filter_parse_handler &&
		child_parse_handler);
```

**Comment by RekGRpth** [comment]
> Maybe we should do the checks in such a way that there are fewer changes?
> 
> ```c
> 	CParseHandlerProperties *prop_parse_handler =
> 		dynamic_cast<CParseHandlerProperties *>((*this)[0]);
> 	CParseHandlerGroupingColList *grouping_col_list_parse_handler =
> 		dynamic_cast<CParseHandlerGroupingColList *>((*this)[1]);
> 	CParseHandlerProjList *proj_list_parse_handler =
> 		dynamic_cast<CParseHandlerProjList *>((*this)[2]);
> 	CParseHandlerFilter *filter_parse_handler =
> 		dynamic_cast<CParseHandlerFilter *>((*this)[3]);
> 	CParseHandlerPhysicalOp *child_parse_handler =
> 		dynamic_cast<CParseHandlerPhysicalOp *>((*this)[4]);
> 	GPOS_ASSERT(prop_parse_handler &&
> 		grouping_col_list_parse_handler &&
> 		proj_list_parse_handler &&
> 		filter_parse_handler &&
> 		child_parse_handler);
> ```

This looks a little strange. I followed this approach: put an assert before using (dereferencing) a pointer.

**Comment by KnightMurloc** [comment]
> This looks a little strange

Why? For me, it is so logical that we get the elements and immediately check them for NULL. I don't think that NULL as a result of casting is the expected behavior in any of the cases.

**Comment by RekGRpth** [comment]
> > This looks a little strange
> 
> Why? For me, it is so logical that we get the elements and immediately check them for NULL. I don't think that NULL as a result of casting is the expected behavior in any of the cases.

Also it is not ORCA style

**Comment by KnightMurloc** [comment]
We can split the checks into separate assert. but in general, the idea of checking immediately after casting seems better to me.

**Comment by RekGRpth** [comment]
> We can split the checks into separate assert. but in general, the idea of checking immediately after casting seems better to me.

Then it looks like
```c
	CParseHandlerProperties *prop_parse_handler =
		dynamic_cast<CParseHandlerProperties *>((*this)[0]);
	GPOS_ASSERT(NULL != prop_parse_handler);
	CParseHandlerGroupingColList *grouping_col_list_parse_handler =
		dynamic_cast<CParseHandlerGroupingColList *>((*this)[1]);
	GPOS_ASSERT(NULL != grouping_col_list_parse_handler);
	CParseHandlerProjList *proj_list_parse_handler =
		dynamic_cast<CParseHandlerProjList *>((*this)[2]);
	GPOS_ASSERT(NULL != proj_list_parse_handler);
	CParseHandlerFilter *filter_parse_handler =
		dynamic_cast<CParseHandlerFilter *>((*this)[3]);
	GPOS_ASSERT(NULL != filter_parse_handler);
	CParseHandlerPhysicalOp *child_parse_handler =
		dynamic_cast<CParseHandlerPhysicalOp *>((*this)[4]);
	GPOS_ASSERT(NULL != child_parse_handler);
```

**Comment by KnightMurloc** [comment]
I meant
```c
	CParseHandlerProperties *prop_parse_handler =
		dynamic_cast<CParseHandlerProperties *>((*this)[0]);
	CParseHandlerGroupingColList *grouping_col_list_parse_handler =
		dynamic_cast<CParseHandlerGroupingColList *>((*this)[1]);
	CParseHandlerProjList *proj_list_parse_handler =
		dynamic_cast<CParseHandlerProjList *>((*this)[2]);
	CParseHandlerFilter *filter_parse_handler =
		dynamic_cast<CParseHandlerFilter *>((*this)[3]);
	CParseHandlerPhysicalOp *child_parse_handler =
		dynamic_cast<CParseHandlerPhysicalOp *>((*this)[4]);
        GPOS_ASSERT(NULL != prop_parse_handler);
        GPOS_ASSERT(NULL != grouping_col_list_parse_handler);
        GPOS_ASSERT(NULL != proj_list_parse_handler);
        GPOS_ASSERT(NULL != filter_parse_handler);
	GPOS_ASSERT(NULL != child_parse_handler);
```

**Comment by RekGRpth** [comment]
> I meant
> 
> ```c
> 	CParseHandlerProperties *prop_parse_handler =
> 		dynamic_cast<CParseHandlerProperties *>((*this)[0]);
> 	CParseHandlerGroupingColList *grouping_col_list_parse_handler =
> 		dynamic_cast<CParseHandlerGroupingColList *>((*this)[1]);
> 	CParseHandlerProjList *proj_list_parse_handler =
> 		dynamic_cast<CParseHandlerProjList *>((*this)[2]);
> 	CParseHandlerFilter *filter_parse_handler =
> 		dynamic_cast<CParseHandlerFilter *>((*this)[3]);
> 	CParseHandlerPhysicalOp *child_parse_handler =
> 		dynamic_cast<CParseHandlerPhysicalOp *>((*this)[4]);
>         GPOS_ASSERT(NULL != prop_parse_handler);
>         GPOS_ASSERT(NULL != grouping_col_list_parse_handler);
>         GPOS_ASSERT(NULL != proj_list_parse_handler);
>         GPOS_ASSERT(NULL != filter_parse_handler);
> 	GPOS_ASSERT(NULL != child_parse_handler);
> ```

I don’t see the logic here; we put the assertion either immediately after casting or before dereferencing.

**Comment by KnightMurloc** [comment]
The point is that in this case we do we don't get wedged into the existing logic and it looks more readable.

**Comment by RekGRpth** [comment]
> The point is that in this case we do we don't get wedged into the existing logic and it looks more readable.

I changed it in the first file to check. It's all right?

**Comment by KnightMurloc** [comment]
Yes. I think it's better

**Comment by RekGRpth** [comment]
> Yes. I think it's better

reordered

**Review by dnskvlnk — APPROVED** [review]

**Review by KnightMurloc — APPROVED** [review]
